R/messages.R

```diff
@@ -110,3 +110,85 @@ setvar <- function(core, dates, var, values, unit)
 
     invisible(NULL)
 }
+
+
+#' Fetch all the available Hector variables
+#'
+#' This function is similar to fetchvars, except it fetches all available
+#' Hector variables, including exogenous inputs (e.g., emissions & pre-industrial
+#' concentrations), parameters (e.g., alpha, beta, ECS), and output
+#' variables (e.g., concentrations & forcings).
+#'
+#' @param core Hector core object
+#' @param dates Vector of dates, optional
+#' @param scenario Scenario name, optional str
+#' @param outpath Absolute path of the output csv. Output will only be written to
+#' csv if this is given
+#' @return vars_all    Dataframe containing all Hector variables
+#' @family main user interface functions
+#' @export
+fetchvars_all <- function(core, dates=NULL, scenario=NULL, outpath=NULL) {
+
+    if (is.null(scenario)) {
+        scenario <- core$name
+    }
+
+    strt <- startdate(core)
+    end <- getdate(core)
+
+    if (is.null(dates)) {
+        dates <- strt:end
+    } else {
+        valid <- dates >= strt & dates <= end
+        dates <- dates[valid]
+    }
+    
+    # Get output for variables that do use the date param
+    rslt_date  <- do.call(rbind,
+                      lapply(vars_date, function(v) {
+                            sendmessage(core, GETDATA(), v, dates, NA, '')
+                      }))
+    
+    # Get variables that DO NOT use date arg
+    rslt_nodate <- do.call(rbind,
+                       lapply(vars_nodate, function(v) {
+                            sendmessage(core, GETDATA(), v, NA, NA, '')
+                       }))
+    
+    rslt <- rbind(rslt_date, rslt_nodate)
+    rslt$scenario <- scenario
+    
+    if (!is.null(outpath)) {
```

Comment:
All of the code from here on is useful enough that I might consider making it its own exported function. I can think of _lots_ of circumstances where we would want to export output in a wide, pretty CSV format. If you wanted to preserve this interface, you could then just call that function here. In other words...

```r
# ...
  if (!is.null(outpath)) {
    write_wide_output(rslt, outpath, ...)
    message("Output written to ", outpath)
  }
  invisible(rslt)
}

# R/write_wide_output.R
#' Write Hector output to a wide CSV file
#' @param results Hector results `data.frame` such as the one produced by [fetchvars()]
#' @param outpath Path for writing output
#' @param ... Additional arguments to [utils::write.table()]
write_wide_output <- function(results, outpath, ...) {
  # ...
  # Note the use of ... below to allow users to modify other arguments of `write.table`
  write.table(rslt_pretty, outpath, col_names = col_names, row.names = FALSE, ...)
}
```